packages/template-retail-react-app/app/components/_app/index.jsx

```diff
@@ -45,37 +53,54 @@ import {
     DEFAULT_SITE_TITLE,
     HOME_HREF,
     THEME_COLOR,
-    CAT_MENU_DEFAULT_NAV_DEPTH,
+    CAT_MENU_DEFAULT_NAV_SSR_DEPTH,
     CAT_MENU_DEFAULT_ROOT_CATEGORY,
     DEFAULT_LOCALE
 } from '../../constants'
 
 import Seo from '../seo'
 import {resolveSiteFromUrl} from '../../utils/site-utils'
-import useMultiSite from '../../hooks/use-multi-site'
-import {useCategory, useCustomerType} from 'commerce-sdk-react-preview'
 
-const App = (props) => {
-    const {children, targetLocale = DEFAULT_LOCALE, messages = {}} = props
-    const {data: allCategories} = useCategory(
-        {parameters: {id: CAT_MENU_DEFAULT_ROOT_CATEGORY, levels: CAT_MENU_DEFAULT_NAV_DEPTH}},
-        {
-            select: (categories) => {
-                // Note: What is the best to handle special case like this?? Should commerce sdk handles this?
-                if (categories.isError) {
-                    const message =
-                        categories.title === 'Unsupported Locale'
-                            ? `
-It looks like the locale “${categories.locale}” isn’t set up, yet. The locale settings in your package.json must match what is enabled in your Business Manager instance.
-Learn more with our localization guide. https://sfdc.co/localization-guide
-`
-                            : categories.detail
-                    throw new Error(message)
-                }
-                return flatten(categories, 'categories')
+const onClient = typeof window !== 'undefined'
+
+/* 
+The categories tree can be really large! For performance reasons,
+we only load the level 0 categories on server side, and load the rest
+on client side to reduce SSR page size.
+*/
+const useLazyLoadCategories = () => {
+    const itemsKey = 'categories'
+
+    const levelZeroCategoriesQuery = useCategory({
+        parameters: {id: CAT_MENU_DEFAULT_ROOT_CATEGORY, levels: CAT_MENU_DEFAULT_NAV_SSR_DEPTH}
+    })
+
+    const ids = levelZeroCategoriesQuery.data?.[itemsKey].map((category) => category.id)
+    const queries = useCategoryBulk(ids, {
+        enabled: onClient && ids?.length > 0
+    })
+    // make sure all queries is successful before returning
+    if (onClient && queries?.every((query) => query.isSuccess)) {
```

Comment:
This is equivalent to `promise.all()` but it doesn't have to be..

Imagine the server is under heavy load and fetching a small category takes 200ms and a large category takes 2000ms. We should be able to show the small category first when response is available. 

```suggestion
    if (onClient)) {
```